scripts/src/javatests/org/oppia/android/scripts/coverage/RunCoverageTest.kt

```diff
@@ -168,173 +193,1169 @@ class RunCoverageTest {
+  }
+
+  @Test
+  fun testRunCoverage_sampleTestsMarkdownFormat_generatesCoverageReport() {
+    val filePathList = listOf("coverage/main/java/com/example/AddNums.kt")
+
+    testBazelWorkspace.initEmptyWorkspace()
+    testBazelWorkspace.addSourceAndTestFileWithContent(
+      filename = "AddNums",
+      testFilename = "AddNumsTest",
+      sourceContent = addSourceContent,
+      testContent = addTestContent,
+      sourceSubpackage = "coverage/main/java/com/example",
+      testSubpackage = "coverage/test/java/com/example"
+    )
+
+    RunCoverage(
+      "${tempFolder.root}",
+      filePathList,
+      ReportFormat.MARKDOWN,
+      longCommandExecutor,
+      scriptBgDispatcher
+    ).execute()
+
+    val outputReportText = File(
+      "${tempFolder.root}" +
+        "$coverageDir/CoverageReport.md"
+    ).readText()
+
+    val expectedResult = getExpectedMarkdownText(filePathList.get(0))
+
+    assertThat(outputReportText).isEqualTo(expectedResult)
+  }
+
+  @Test
+  fun testRunCoverage_withMultipleFilesMarkdownFormat_generatesCoverageReport() {
+    val oppiaDevelopGitHubLink = "https://github.com/oppia/oppia-android/tree/develop"
+    val filePathList = listOf(
+      "coverage/main/java/com/example/AddNums.kt",
+      "coverage/main/java/com/example/SubNums.kt"
+    )
+
+    val subSourceContent =
+      """
+      package com.example
+      
+      class SubNums {
+        companion object {
+          fun subNumbers(a: Int, b: Int): Any {
+            return if (a == 0 && b == 0) {
+                "Both numbers are zero"
+            } else {
+                a - b
+            }
+          }
+        }
+      }
+      """.trimIndent()
+
+    val subTestContent =
+      """
+      package com.example
+      
+      import org.junit.Assert.assertEquals
+      import org.junit.Test
+      
+      class SubNumsTest {
+        @Test
+        fun testSubNumbers() {
+          assertEquals(SubNums.subNumbers(1, 0), 1)
+          assertEquals(SubNums.subNumbers(4, 3), 1)         
+        }
+      }
+      """.trimIndent()
+
+    testBazelWorkspace.initEmptyWorkspace()
+
+    testBazelWorkspace.addSourceAndTestFileWithContent(
+      filename = "AddNums",
+      testFilename = "AddNumsTest",
+      sourceContent = addSourceContent,
+      testContent = addTestContent,
+      sourceSubpackage = "coverage/main/java/com/example",
+      testSubpackage = "coverage/test/java/com/example"
+    )
+
+    testBazelWorkspace.addSourceAndTestFileWithContent(
+      filename = "SubNums",
+      testFilename = "SubNumsTest",
+      sourceContent = subSourceContent,
+      testContent = subTestContent,
+      sourceSubpackage = "coverage/main/java/com/example",
+      testSubpackage = "coverage/test/java/com/example"
+    )
+
+    RunCoverage(
+      "${tempFolder.root}",
+      filePathList,
+      ReportFormat.MARKDOWN,
+      longCommandExecutor,
+      scriptBgDispatcher
+    ).execute()
+
+    for (file in filePathList) {
```

Comment:
Here & below (I think in one other test): prefer directly asserting on exact state rather than iterating the list. Control flow in tests is difficult to verify.

I think you'd probably benefit from leveraging some helper logic here, too--a lot of the tests are doing redundant checks (we generally want to keep tests as specific as possible to verifying whatever should fail if their conditions are broken and rely on other tests to verify related behavior rather than checking the same things many times, though this isn't always feasible).

Reply:
### Status: Yet to do
---
Ah makes sense. Few places I can think I can utilize helper logic would be to get the test and source content since they're being initialized redundantly in many places, already have helper functions for acquiring common html and md reports and will try to add any other wherever possible and I will also try if possible to may be have/split the existing test cases to be more specific like these ones I guess,
- testRunCoverage_withMultipleFiles_generatesOverallCoveragePercentage ->
```
...
assertThat(outputReportText).contains("Overall Coverage: **62.50%**")
```
- testRunCoverage_withProtoReportFormat_skipsCoverageGeneration
```
...
// Since report Format is PROTO no report generation would have been initiated
// resulting in no coverage report directory or files being created.
assertThat(directory.isDirectory).isFalse()
```